### ISSUE TITLE
Feature Request: OS Detection

### ISSUE THREAD

**Issue by lyallcooper** [zmkfirmware/zmk#2553]
It would be nice if zmk offered a host OS detection feature (see [QMK for example](https://docs.qmk.fm/features/os_detection)) to allow dynamic binding of keys or behavior depending on the OS of the host the keyboard is connected to.

My specific use case: I have a keyboard that I use with macOS and Windows. I have a layer where the modifier keys are in the normal Mac order (ctrl, opt, cmd) and a layer where the mods are in the normal Windows order (ctrl, win, alt). I currently default to the Mac layer, but manually switch to the Windows one whenever I connect the keyboard to the Windows machine. It would be nice if the keyboard could detect this on startup and set the appropriate keys on its own.

**Comment by btilford** [comment]
This would be useful for those of us that are forced to use a Mac at work and are constantly having issues with shortcuts like copy paste. With QMK I was able build out some macros that were cross platform.

**Comment by eb0687** [comment]
Is this feature part of the roadmap or not really?

**Comment by caksoylar** [comment]
I would guess not, but a PoC implementation (especially one with BLE) could help with any movement in this area.

**Comment by peterjc** [comment]
This Reddit post and thread describes the origin of the QMK OS-detection code:

https://www.reddit.com/r/olkb/comments/x1ezbg/way_to_detect_host_os_in_qmk/

Key quote from author @KapJI:

> It uses wLength field in get_descriptor packets during USB setup sequence. Different OSes have different number of such packets and some patterns in these values. I guess it depends on the implementation of USB stack. For example, Linux always sets it to 0xFF. 

It was inspired by the idea of looking at descriptor lengths as in the minimal proof-of-principle MIT licensed source https://github.com/keyboardio/FingerprintUSBHost/blob/master/src/FingerprintUSBHost.cpp - but all the logic there seems to be a way to spot Linux which uses a fixed descriptor request length of 255 or 0xFF.

The code in QMK is much more sophisticated, looking at the number of descriptors and their lengths. There is a set of test data in the orginal code with lists of lengths from Windows 10 (long), macOS, iOS/iPadOS, assorted Linux (always just three lengths: 255, 255, 255), PS5, Nintendo Switch, and Quest 2.

Now, given QMK is GPL licensed, ZMK has a very clear clean room policy: https://zmk.dev/docs/development/contributing/clean-room

So, @KapJI, would you be kind enough to grant permission to re-use that set of length data under the MIT license (so that we can use it to derive a new set of heuristics independently)?

Or, going further, would you be willing to allow for your code and algorithm for OS detection as contributed to QMK to be used under the MIT license (in which case we can ask the other QMK contributors over the incremental improvements since)?

**Comment by KapJI** [comment]
@peterjc Yes, I'm the original author of the OS detection feature contributed to QMK.

I'm happy for my original contributions (the descriptor length dataset and the OS detection algorithm) to be reused under the MIT license as well as GPLv2. This only applies to the parts written by me — if there were later improvements from other QMK contributors they would need to approve those separately.

Also, it would be great if the dataset could be expanded with observations from newer systems. And if you end up collecting additional data or refining the heuristics, it would be nice if those findings could be shared back with QMK so both projects benefit.

**Comment by peterjc** [comment]
@KapJI thank you! And yes, needing permission from the authors of later changes to the QMK to use their work is clear.

Your initial dataset had [this as a code comment](https://github.com/KapJI/qmk_firmware/commit/fece38f60388c0d529f11d2952fa11876d7024c8#diff-2240906bfa7fc7108dac07c13af8f3679bb91474478bedb33bf0701e5ab66313R30-R38):

```
// Some data I collected:
// Windows 10: [FF, FF, 4, 24, 4, 24, 4, FF, 24, FF, 4, FF, 24, 4, 24, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A, 20A]
// Another host with Windows 10: [FF, FF, 4, 24, 4, 24, 4, 24, 4, 24, 4, 24]
// macOS 12.5: [2, 24, 2, 28, FF]
// iOS/iPadOS 15.6: [2, 24, 2, 28]
// Linux (including Android, Raspberry Pi and WebOS TV): [FF, FF, FF]
// PS5: [2, 4, 2, 28, 2, 24]
// Nintendo Switch: [82, FF, 40, 40, FF, 40, 40, FF, 40, 40, FF, 40, 40, FF, 40, 40]
// Quest 2: [FF, FF, FF, FE, FF, FE, FF, FE, FF, FE, FF]
```

**Comment by peterjc** [comment]
I initally got something rather different from an ARM Mac Mini, running Sequoia 15.7.4, but now it gives the same at my Intel Mac's data (see below)... the initial values turned out to be random data prior to saving a finger print.

---

This is the QMK logging from my Nintendo Switch, `[82, FF, 40, 40, 40, FF, 40, 40, 40, FF, 40, 40, 40]`:

<details>

```
HID console connected: peterjc Bivouac34 (FEED:0000:0100)
> i: 0, wLength: 0x82
> i: 1, wLength: 0xFF
> i: 2, wLength: 0x40
> i: 3, wLength: 0x40
> i: 4, wLength: 0x40
> i: 5, wLength: 0xFF
> i: 6, wLength: 0x40
> i: 7, wLength: 0x40
> i: 8, wLength: 0x40
> i: 9, wLength: 0xFF
> i: 10, wLength: 0x40
> i: 11, wLength: 0x40
> i: 12, wLength: 0x40
```

</details>

Similar to but not quite the same as KapJI's notes:

---

My older Intel iMac, running Monterey 12.7.6, no USB hub involved, `[02, 14, 02, 10, 02, 42]`:

<details>

```
HID console connected: peterjc Bivouac34 (FEED:0000:0100)
> i: 0, wLength: 0x02
> i: 1, wLength: 0x14
> i: 2, wLength: 0x02
> i: 3, wLength: 0x10
> i: 4, wLength: 0x02
> i: 5, wLength: 0x42
```

</details>

Same on ARM Mac Mini, running Sequoia 15.7.4.

Again similar to but different to the example for macOS 12.5 in KapJI's notes.

---

Windows 11, `[FF, FF, FF, 4, 14, 4, 14, 4, 14, 1F4, FF, FF, FF, FF, FF, FF, FF, FF, FF, FF, FF, FF]`, no string of 20A like KapJI reported for Windows 10, but still several 4 and even more FF.

<details>

```
HID console connected: peterjc Bivouac34 (FEED:0000:0100)
> i: 0, wLength: 0xFF
> i: 1, wLength: 0xFF
> i: 2, wLength: 0xFF
> i: 3, wLength: 0x04
> i: 4, wLength: 0x14
> i: 5, wLength: 0x04
> i: 6, wLength: 0x14
> i: 7, wLength: 0x04
> i: 8, wLength: 0x14
> i: 9, wLength: 0x1F4
> i: 10, wLength: 0xFF
> i: 11, wLength: 0xFF
> i: 12, wLength: 0xFF
> i: 13, wLength: 0xFF
> i: 14, wLength: 0xFF
> i: 15, wLength: 0xFF
> i: 16, wLength: 0xFF
> i: 17, wLength: 0xFF
> i: 18, wLength: 0xFF
> i: 19, wLength: 0xFF
> i: 20, wLength: 0xFF
> i: 21, wLength: 0xFF
```

</details>

**Comment by peterjc** [comment]
In conversation with @Drashna on the QMK Discord, who is one of the QMK contributors to this area of code, they noted there have been multiple one-off contributors, so dual licensing any of the current QMK code isn't practical. We agreed given ZMK has permission to use @JapKI's original `wLength` examples and heuristics under the MIT License, and current `wLength` examples, a ZMK contributor should be able to do a clean room implentation without referring to current QMK.

(Not me, as I have looked at this area of QMK too much.)

**Comment by peterjc** [comment]
There's some AI assisted work at https://github.com/cormoran/zmk-feature-os-detection with some detailed fingerprinting logs at https://github.com/cormoran/zmk-feature-os-detection/blob/main/docs/fingerprints.md covering both USB (starting from the basic idea as discussed above used in QMK) and BLE (what GATT requests are received).
